_posts/2022-05-31-introduction-to-project-panama-part-1.markdown

```diff
@@ -121,86 +119,98 @@ FunctionDescriptor printfDescriptor = FunctionDescriptor.of(
 
-So, at this stage, I have introduced all necessary terms to explain what both down- and upcalls are, in terms of Java -
-_downcall_ is an invocation of a native function through a **MethodHandle** formed from a native function address and its Java version of a function descriptor.
-At the same time, the _upcall_ is an invocation of some code written in Java through a **MethodHandle** converted
-into a native memory segment which could then be passed to native functions as a function pointer.
+With all necessary concepts explained, it would be better to extend both definitions of down- and upcalls:
+* a _downcall_ is an invocation of a native function through a **MethodHandle** formed from a native function address and its Java version of a function descriptor;
+* an _upcall_ is an invocation of some code written in Java through a **MethodHandle** converted into a native memory 
+segment which could then be passed to native functions as a function pointer.
 
 
-### Task: allocate a native memory.
+### 4. Allocate native memory.
 
 Okay, at this stage you may think we have pretty much everything we need to perform the actual downcall, but that's not all.
 We need somehow to bind Java objects to native memory segments to make sure C _printf_ can access them.
-Here’s the moment when we step into the area of **Foreign Memory Access API** (FMA)!
 
-Way back in time when I did C programming (in high school, honestly), a memory allocation and freeing were kinda painful
-because no matter how good you are you’d, eventually you will forget to allocate or free a memory which will cause a
+Both memory allocation and freeing were painful because no matter how good software developer you are, eventually you will forget to allocate or free a memory which will cause a
 program to leak it or blow up with a segmentation fault.
 
-In Java, we traditionally relied on the GC to take care of memory cleaning up.
+Java relies on a Garbage Collector to allocate and free memory but Panama’s FMA API allocates memory off-heap.
 In Panama, the FMA API helps to allocate a memory off-heap which is a crucial part of any native interop story!
 
 The FMA API allows to allocate and access memory segments, their addresses, and the shape of contiguous memory
 regions, located either on- or off- the Java heap.
-In concepts of Panama, all allocated memory segments must be bound to a specific memory session (**MemorySession**),
-in other terms, consider a session like a unified memory allocation tool,
-for instance, C `malloc`, but auto-closable (thanks to Java try-with-resources syntax block).
+All allocated memory segments are bound to a specific memory session (**MemorySession**), 
+an instance of it offers a set of APIs to allocate memory segments. Consider a memory session like a unified memory allocation tool,
+like C `malloc`.
+One of the important features of **MemorySession** is an implementation of an **AutoClosable** interface - 
```

Comment:
**MemorySession** implements the **AutoClosable** interface which greatly simplifies de-allocation using the try-with-resources construct.